ports/stm32/flash.c: WB55 flash locking.

jimmo opened the pull request:
(Extends from #6594, just the last three commits are relevant to this PR)

The WB55 wireless core shares the same internal flash as the app core, and so when the app core does flash erase or write operations, then the wireless core will stall as the bus is locked.

This adds a BLE multitest to demonstrate the issue, and then fixes it by:
 - Processing HCI data as quickly as possible in the RX IRQ (this requires that we no longer support btstack on WB55 for now)
 - Implementing the flash locking algorithm described in AN5289 Rev 3, Figure 10.

pi-anl committed:
ports/unix: Handle pending events/scheduler in MICROPY_EVENT_POLL_HOOK.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Re-instate optional no-ringbuf modbluetooth.

This requires that the event handlers are called from non-interrupt context
(i.e. the scheduler).

This will allow the BLE stack (e.g. NimBLE) to run from the scheduler rather
than PENDSV and invoke Python callbacks directly.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Make UUID type accessible outside modbluetooth.c.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/run-multitests.py: Fix diff order show changes in test relative to truth.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/multi_bluetooth: Improve reliability of event waiting.

Use the same `wait_for_event` in all tests that doesn't hold a reference
to the event data tuple and handles repeat events.

Also fix a few misc reliability issues around timeouts and sequencing.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/run-multitests.py: Add a -p flag to run permutations of instances.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Make "data available split" across multiple events.

Instead of having the stack indicate a "start", "data"..., "end",
just send "data" and let the Python handler sort it out with a new
"offset" field in the data tuple.

This is because the non-ringbuffer modbluetooth implementation can't do
the re-assembly.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/nimble: Poll startup directly rather than using NimBLE sem.

Using a semaphore (the previous approach) will only run the UART while,
whereas for startup we need to also run the event queue.

This change makes it run the full scheduler hook.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
ports/unix: Make mp_hal_delay_ms run MICROPY_EVENT_POLL_HOOK.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/nimble: Make STM32 & Unix nimble ports use the scheduler.

This changes STM32 from using PENDSV to run NimBLE to the MicroPython
scheduler instead. This allows Python callbacks to be invoked directly
(and therefore synchronously) rather than via the ringbuffer.

The NimBLE UART HCI and event processing now happens in a scheduled task every 128ms. When RX IRQ events arrive, it will also schedule this task to improve latency.

Similar change for Unix where the background thread now queues the scheduled task.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/multi_bluetooth: Add a test for WB55 concurrent flash access.

This test currently passes on Unix/PYBD, but fails on WB55 because
it lacks synchronisation of the internal flash.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
ports/stm32/rfcore.c: Make RX IRQ schedule the NimBLE handler.

jimmo committed:
ports/stm32/flash.c: WB55 flash locking.

This is needed to moderate concurrent access to the internal flash,
as while an erase/write is in progress, execution will stall on
the wireless core due to the bus being locked.

This implements Figure 10 from AN5289 Rev 3.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
ports/stm32/rfcore.c: Don't send HCI ACL cmds while another is pending.

dpgeorge commented on tests/multi_bluetooth/stress_log_filesystem.py:
```diff
@@ -0,0 +1,185 @@
+# Test BLE GAP connect/disconnect
```
this comment needs to be changed (cut and paste error?)

dpgeorge commented on tests/multi_bluetooth/stress_log_filesystem.py:
```diff
@@ -0,0 +1,185 @@
+        # conn_handle, start_handle, end_handle, uuid = data
+        if data[-1] == SERVICE_UUID:
+            print("_IRQ_GATTC_SERVICE_RESULT", data[3])
+            waiting_events[event] = (data[1], data[2],)
```
please remove trailing comma on RHS, and then run `black -l99` on this file (to fix other formatting issues)

dpgeorge commented on tests/multi_bluetooth/stress_log_filesystem.py:
```diff
@@ -0,0 +1,185 @@
+        periodic_log_write()
+        if event in waiting_events:
+            result = waiting_events[event]
+            del waiting_events[event]
```
I think `waiting_events.pop(event)` is a better way to write this, it's simpler and more atomic (especially if IRQs can change this dict).

Could even write this as:
```python
if result := waiting_events.pop(event, -1) != -1:
    return result
```
But not sure if it's good to put a dependency on the walrus operator....

dpgeorge commented on tests/multi_bluetooth/stress_log_filesystem.py:
```diff
@@ -0,0 +1,185 @@
+            wait_for_event(_IRQ_CENTRAL_DISCONNECT, 2*TIMEOUT_MS)
+    finally:
+        ble.active(0)
+        log_file.close()
```
would it make sense for the test to re-open the file and print out its contents at the end here, for further verification it ran correctly?

jimmo replied:
This file will not be exactly the same across runs.

dpgeorge replied:
Ok

dpgeorge commented on ports/stm32/rfcore.c:
```diff
@@ -597,7 +600,8 @@ void rfcore_ble_hci_cmd(size_t len, const uint8_t *src) {
 void rfcore_ble_check_msg(int (*cb)(void *, const uint8_t *, size_t), void *env) {
     parse_hci_info_t parse = { cb, env, false };
-    tl_check_msg_ble(&ipcc_mem_ble_evt_queue, &parse);
+    // Either this is being called
```
not sure what this comment means...?

dpgeorge commented on ports/stm32/rfcore.c:
```diff
@@ -556,11 +555,15 @@ void rfcore_ble_init(void) {
     // Clear any outstanding messages from ipcc_init
     tl_check_msg(&ipcc_mem_sys_queue, IPCC_CH_SYS, NULL);
-    tl_check_msg_ble(&ipcc_mem_ble_evt_queue, NULL);
+    tl_check_msg(&ipcc_mem_ble_evt_queue, IPCC_CH_BLE, NULL);
```
is this line even needed anymore, now that events are handled immediately in the RX IRQ?

dpgeorge commented:
Now that RX IRQs are executed immediately, it could be that IPCC IRQ calls `mp_bluetooth_on_hci_irqs` calls `mp_bluetooth_hci_poll_uart` calls `mp_bluetooth_nimble_hci_uart_process` calls into NimBLE, even if the BLE stack (nimble) is disabled.... does this matter, does it need a guard?

jimmo committed:
tests/multi_bluetooth: Change del to pop to clear event.

jimmo commented:
Rebased on latest updates to #6594. We may need to introduce an RX buffer for rfcore.c as there could be some delay between the IRQ arriving and the scheduler executing and it might be possible for the payload to be replaced before being processed. However with the current changes the tests (including the new one) pass fine, so it's already better than what was there before so that might need to be a future improvement. Also added a fix for sending large payloads where NimBLE sends multiple HCI payloads before receiving acks for the previous ones leading to lost packets.

> Now that RX IRQs are executed immediately, it could be that IPCC IRQ calls `mp_bluetooth_on_hci_irqs` calls `mp_bluetooth_hci_poll_uart` calls `mp_bluetooth_nimble_hci_uart_process` calls into NimBLE, even if the BLE stack (nimble) is disabled.... does this matter, does it need a guard?

No longer required because NimBLE now runs entirely in the scheduler.

dpgeorge commented:
This was rebased and merged in 7e75245d549d01ac4f01ee615c193c00cf3269a7 through 21c293fbcd1f9d5bafac300b5da265b4efd2be52

Thank you!
